polaris/dataset/zarr/_checksum.py

```diff
@@ -0,0 +1,314 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import hashlib
+import heapq
+import os
+import re
+from dataclasses import asdict, dataclass, field
+from functools import total_ordering
+from json import dumps
+from pathlib import Path
+from typing import Optional
+
+import datamol as dm
+import zarr
+import zarr.errors
+from fsspec import AbstractFileSystem
+from tqdm import tqdm
+
+from polaris.utils.errors import InvalidZarrChecksum
+
+ZARR_DIGEST_PATTERN = "([0-9a-f]{32})-([0-9]+)--([0-9]+)"
+
+
+def compute_zarr_checksum(zarr_root_path: str, fs: Optional[AbstractFileSystem] = None) -> str:
+    r"""
+    Implements an algorithm to compute the Zarr checksum.
+
+    Warning: This checksum is sensitive to Zarr configuration. 
+        This checksum is not insensitive to change in the Zarr structure. For example, if you change the chunk size, 
+        the checksum will also change.
+
+    To understand how this works, consider the following directory structure:
+
+               . (root)
+              / \
+             a   c
+            /
+           b
+    
+    Within zarr, this would for example be:
+
+    - `root`: A Zarr Group with a single Array.
+    - `a`: A Zarr Array
+    - `b`: A single chunk of the Zarr Array
+    - `c`: A metadata file (i.e. .zarray, .zattrs or .zgroup) 
+
+    To compute the checksum, we first find all the trees in the node, in this case b and c. 
+    We compute the hash of the content (the raw bytes) for each of these files.
+    
+    We then work our way up the tree. For any node (directory), we find all children of that node.
+    In an sorted order, we then serialize a list with - for each of the children - the checksum, size, and number of children.
+    The hash of the directory is then equal to the hash of the serialized JSON.
+
+    The Polaris implementation is heavily based on the [`zarr-checksum` package](https://github.com/dandi/zarr_checksum).
+    This method is the biggest deviation of the original code.
+    """
+
+    if fs is None:
+        # Try guess the filesystem if it's not specified
+        fs = dm.utils.fs.get_mapper(zarr_root_path).fs
+
+    # Get the protocol of the path
+    protocol = dm.utils.fs.get_protocol(zarr_root_path, fs)
+
+    # For a local path, we extend the path to an absolute path
+    # Otherwise, we assume the path is already absolute
+    if protocol == "file":
+        zarr_root_path = os.path.expandvars(zarr_root_path)
+        zarr_root_path = os.path.expanduser(zarr_root_path)
+        zarr_root_path = os.path.abspath(zarr_root_path)
+
+    # Make sure the path exists and is a Zarr archive
+    zarr.open_group(zarr_root_path, mode="r")
+
+    # Generate the checksum
+    tree = ZarrChecksumTree()
+
+    # Find all files in the root
+    leaves = fs.find(zarr_root_path, detail=True)
+
+    for file in tqdm(leaves.values(), desc="Finding all files in the Zarr archive"):
```

Comment:
With up- and downloads, it is a little harder because this is currently handled by Zarr (see [`copy_all`](https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.copy_all) and [`copy_store`](https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.copy_store)). These functions expose the `log` argument to log updates on the upload, but I am not a big fan of the formatting. Might nevertheless be a good idea to start using that for a better UX.